Architecture.md

```diff
@@ -0,0 +1,100 @@
+
+Because a single polymorphic record type may be instantiated to different concrete types at different places in the AST, we must also inline while we monomorphize. 
+
+The general monomorphization procedure is as follows (not a full algorithm): 
+
+  1. We traverse the AST until we reach an `App` node. 
+  2. We "peel" the App node and get an `(f,args)` where `f` is the "terminal" function expression and `args` is a list of the argument expressions to which it is applied. 
+  3. We check whether the type of `f` is already monomorphic. If it is, we make no change to the node. 
+  4. If `f` is not monomorphic, we strip the quantifiers and check whether we can specialize any bound type variables to a concrete type. E.g. if `f :: forall x. Tuple x x -> x -> Int` and the arguments are `[Tuple String String, String]`, we instantiate `x` to `String`. 
+  5. We recurse down the list of arguments. If we encounter an argument that is a free (term-level) variable that is *not* a Plutus builtin, we inline it, monomorphizing it to the concrete type of the argument if possible.  
+  6. We apply the (possibly monomorphized & inlined) argument to the function and procede to the next argument 
+     until we run out of arguments to process.
+  7. Finally, we re-quantify the type of `f` in the resulting expression. (It may contain free type variables of kind `Type`, which can be represented in PIR)
+    
+The procedure for inlining is: 
+  1. We check whether the (term level) variable we aim to inline is locally scoped (i.e. qualified by source position) or globally scoped (e.g. qualified by module name). For now, we only attempt to inline globally scoped variables (...I forget why...)
+  2. If the variable is globally scoped (i.e. is the name of a module-level declaration), we lookup the body of the declaration. There are two possibilities here: Either the declaration will be a single non-recursive binding, or it will be a member of a mutually recursive binding group. (PS compiler sorts these for us) 
+    - If the declaration is non-recursive we "walk" its expression-body and monomorphize/inline the sub-expressions as necessary in order to properly assign the expression the type that it is to be monomorphized to. 
+    - If the declaration the member of a recursive binding group, we pause inlining, walk the expression, and "collect" a `Map Name (Name,SourceType,Expr)` where the key is the original name of the expression, and the values are, respectively: The new "fresh" name to give to the monomorphized expression, the monomorphic type that we must assign the expression to, and the body of the declaration. We do this recursively until we have collected every member of the recursive binding group used in the target expression. Finally, we use that map to construct a *monomorphic* mutually recursive binding group (where the names are all fresh) and create a `Let`-binding for the monomorphized mutually recursive group. 
```

Comment:
Nope, this is super outdated. I should probably update this rather than delete it, but I'll flag that as something to do later as it's not essential right now. (The comments in the modules that explain things ought to be correct)